zebra-network/src/meta_addr.rs

```diff
@@ -17,55 +19,314 @@ use zebra_chain::serialization::{
-/// liveness based on the current time. This derived state is tracked using
-/// [`AddressBook::maybe_connected_peers`] and
-/// [`AddressBook::reconnection_peers`].
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+/// Zebra also tracks how recently we've had different kinds of interactions with
+/// each peer, and derives peer usage based on the current time. This derived state
+/// is tracked using [`AddressBook::recently_used_peers`] and
+/// [`AddressBook::candidate_peers`].
+///
+/// To avoid depending on untrusted or default data, Zebra tracks the required
+/// and optional data in each state. State updates are applied using
+/// [`MetaAddrChange`]s.
+///
+/// See the [`CandidateSet`] for a detailed peer state diagram.
+#[derive(Copy, Clone, Debug)]
 #[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
 pub enum PeerAddrState {
-    /// The peer has sent us a valid message.
+    /// The peer's address has just been fetched from a DNS seeder.
     ///
-    /// Peers remain in this state, even if they stop responding to requests.
-    /// (Peer liveness is derived from the `last_seen` timestamp, and the current
-    /// time.)
-    Responded,
+    /// Zebra attempts to connect to all DNS seeder peers on startup. So the
+    /// initial peer connector updates all DNS seeder peers to `AttemptPending`.
+    ///
+    /// If any DNS seeder addresses remain, they are attempted after disconnected
+    /// `Responded` peers.
+    NeverAttemptedDnsSeeder,
 
-    /// The peer's address has just been fetched from a DNS seeder, or via peer
-    /// gossip, but we haven't attempted to connect to it yet.
-    NeverAttemptedGossiped,
+    /// The peer's address has just been fetched via peer gossip, but we haven't
+    /// attempted to connect to it yet.
+    ///
+    /// Gossiped addresses are attempted after DNS seeder addresses.
+    NeverAttemptedGossiped {
+        /// The time that another node claims to have connected to this peer.
+        /// See `get_untrusted_last_seen` for details.
+        #[cfg_attr(
+            any(test, feature = "proptest-impl"),
+            proptest(strategy = "datetime_u32()")
+        )]
+        untrusted_last_seen: DateTime<Utc>,
+        /// The services another node claims this peer advertises.
+        /// See `get_services` for details.
+        untrusted_services: PeerServices,
+    },
 
     /// The peer's address has just been received as part of a `Version` message,
     /// so we might already be connected to this peer.
     ///
     /// Alternate addresses are attempted after gossiped addresses.
-    NeverAttemptedAlternate,
+    NeverAttemptedAlternate {
+        /// The last time another node gave us this address as their canonical
+        /// address.
+        /// See `get_untrusted_last_seen` for details.
+        #[cfg_attr(
+            any(test, feature = "proptest-impl"),
+            proptest(strategy = "datetime_u32()")
```

Comment:
This is `Utc::now()`, not a serialized date:
```suggestion
            proptest(strategy = "datetime_full()")
```